### PR TITLE
Nonblocking master service

### PR DESCRIPTION
**Pull request by DaveCTurner**
The changes introduced in #92021 mean that there is no need for the master service to block its thread while waiting for each publication to complete. This commit removes the now-unnecessary blocking.

This commit also removes the now-unnecessary fake blocking in the `FakeThreadPoolMasterService` used in tests, bringing the implementation covered by tests closer to the production implementation.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java [hunk 1]
```diff
@@ -2369,7 +2377,7 @@ public void taskSucceeded(ClusterStateUpdateTask clusterStateTaskListener, Void
                     public ClusterState execute(ClusterState currentState) {
                         assertEquals(priority, prioritiesQueue.poll());
                         assertEquals(priority, priority());
-                        return currentState;
+                        return randomBoolean() ? currentState : ClusterState.builder(currentState).build();
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java:2380]: We're using a single-threaded executor here, so by sometimes triggering a publication and sometimes forking the publication handling onto a separate thread we demonstrate that there is no longer any blocking involved in the publication process.

**Review by DaveCTurner — COMMENTED** [review]
tip: ignoring whitespace will make this easier to review

**Review by henningandersen — COMMENTED** [review]
I think this looks good, but have a question/comment that you can maybe clarify quickly?

File: server/src/main/java/org/elasticsearch/cluster/service/MasterService.java [hunk 2]
```diff
@@ -413,17 +440,15 @@ protected void publish(
...
-        clusterStatePublisher.publish(clusterStatePublicationEvent, fut, ackListener);
-
-        ActionListener.completeWith(
-            publicationListener,
-            () -> FutureUtils.get(fut) // indefinitely wait for publication to complete
+        clusterStatePublisher.publish(
+            clusterStatePublicationEvent,
+            // fork the completion of publicationListener back onto the master service thread, mainly for legacy reasons; note that this
+            // might be rejected if the MasterService shut down mid-publication
+            new ThreadedActionListener<>(
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/service/MasterService.java:447]: This relies on the single-threaded nature of `forkQueueProcessor`: by the time we get here there is nothing else in the executor's queue, and we won't enqueue anything else there until we complete `publicationListener`.

[the remark this answers is not in the document]
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/cluster/service/MasterService.java:447]: Thanks, that simplifies the review of this.

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/service/MasterService.java:449]: np; I added a little extra commentary in 1530281d56e4ce58bc5582ce7a07aca21889632e.

**Comment by DaveCTurner** [comment]
Bah, ok, we're not always on the master thread where I thought we were (because rejections might leave us on the applier thread). I'll remove that comment for now and follow up later.

File: server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java [hunk 3]
```diff
@@ -176,7 +176,15 @@ protected ExecutorService createThreadPoolExecutor() {
         masterService.setClusterStatePublisher((clusterStatePublicationEvent, publishListener, ackListener) -> {
             clusterStateRef.set(clusterStatePublicationEvent.getNewState());
             ClusterServiceUtils.setAllElapsedMillis(clusterStatePublicationEvent);
-            publishListener.onResponse(null);
+            threadPool.executor(randomFrom(ThreadPool.Names.SAME, ThreadPool.Names.GENERIC)).execute(() -> {
+                final var threadName = Thread.currentThread().getName();
+                try {
+                    Thread.currentThread().setName("[" + MasterService.MASTER_UPDATE_THREAD_NAME + "]");
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java:182]: This seems unnecessary now that we use the `ThreadedActionListener`. I think we should remove this line (and the try-finally), that would validate that we can call the publish listener from any thread.

**Review by henningandersen — APPROVED** [review]
LGTM.

File: server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/service/MasterServiceTests.java]: Thanks, yes - see 464358cb7360de13b64cbf48f0f646069540be4a.

**Pull request merged**